_getline.c

```diff
@@ -0,0 +1,106 @@
+#include "shell.h"
+
+/**
+ * _getline - read a line from a file descriptor and allocate memory for it.
+ * @fd: the file descriptor from which to read the line.
+ * @line: a pointer to a char pointer where the resulting line will be stored.
+ *
+ * This function reads a line from the specified file descriptor and allocates
+ * memory to store the line. The memory for the line should be freed by the
+ * caller when it is no longer needed.
+ *
+ * Return: Upon success, the function returns the number of characters read,
+ * excluding the null-terminator. On end of file or error, it returns -1.
+ */
+
+int _getline(char **line, const int fd)
+{
+	static char *rd = NULL;
+	char buffer[SIZE];
+	int pos;
+
+	pos = _strchr(rd, '\n');
+
+	if (pos >= 0)
+	{
+		process_line(&*line, &rd, pos);
+		return (pos);
+	}
+
+	while ((pos = _read(fd, &rd, buffer)) >= 0)
```

Comment:
btw I got confused a bit using the same name read preceded with an underscore is tricky.
we proceed with a function name with an underscore for a re-implemented function by us but still do the same job as the original functions. please try to come up with a new name for the function `_read`.